File: RailgunVB/Core/Configuration/DiscordConfig.vb
Imports RailgunVB.Core.Logging

Namespace Core.Configuration
    
    Public Class DiscordConfig
    
        Public ReadOnly Property Token As String
        Public Property Prefix As String
        Public ReadOnly Property MasterAdminId As ULong
        Public ReadOnly Property MasterGuildId As ULong = 0
        Public ReadOnly Property BotLogChannels As BotLogChannels = New BotLogChannels
        Public ReadOnly Property OtherAdmins As List(Of ULong) = New List(Of ULong)()
        
        Public Sub New(token As String, prefix As String, masterAdminId As ULong)
            Me.Token = token
            Me.Prefix = prefix
            Me.MasterAdminId = masterAdminId
        End Sub
        
        Public Sub AssignMasterGuild(guildId As ULong)
            _MasterGuildId = guildId
        End Sub
        
        Public Sub AssignBotLogChannel(channelId As ULong, botLogType As BotLogType)
            Select botLogType
                Case BotLogType.Common:
                    _BotLogChannels.Common = channelId
                    Exit Select
                Case BotLogType.CommandManager:
                    _BotLogChannels.CommandMngr = channelId
                    Exit Select
                Case BotLogType.GuildManager:
                    _BotLogChannels.GuildMngr = channelId
                    Exit Select
                Case BotLogType.MusicManager:
                    _BotLogChannels.MusicMngr = channelId
                    Exit Select
                Case BotLogType.MusicPlayer:
                    _BotLogChannels.MusicPlayer = channelId
                    Exit Select
                Case BotLogType.AudioChord:
                    _BotLogChannels.AudioChord = channelId
                    Exit Select
                Case BotLogType.TaskScheduler:
                    _BotLogChannels.TaskSch = channelId
                    Exit Select
                Case BotLogType.TimerManager:
                    _BotLogChannels.TimerMngr = channelId
                    Exit Select
            End Select
        End Function
        
        Public Function AssignAdmin(userId As ULong) As Boolean
            If _OtherAdmins.Contains(userId) Then Return False
            
            _OtherAdmins.Add(userId)
            Return True
        End Function
        
        Public Function RemoveAdmin(userId As ULong) As Boolean
            If Not (_OtherAdmins.Contains(userId)) Then Return False
            
            _OtherAdmins.Remove(userId)
            Return True
        End Function
        
    End Class
    
End NameSpace

File: Railgun/Core/Configuration/DiscordConfig.cs
using System.Collections.Generic;
using Newtonsoft.Json;
using Railgun.Core.Enums;

namespace Railgun.Core.Configuration
{
    [JsonObject(MemberSerialization.Fields)]
    public class DiscordConfig
    {
        public string Token { get; }
        public string Prefix { get; set; }
        public ulong MasterAdminId { get; }
        public ulong MasterGuildId { get; set; } = 0;
        public List<ulong> OtherAdmins { get; } = new List<ulong>();
        public BotLogChannels BotLogChannels { get; } = new BotLogChannels();

        public DiscordConfig(string token, string prefix, ulong masterAdminId) {
            Token = token;
            Prefix = prefix;
            MasterAdminId = masterAdminId;
        }

        [JsonConstructor]
        private DiscordConfig(string token, string prefix, ulong masterAdminId, ulong masterGuildId, BotLogChannels botLogChannels, List<ulong> otherAdmins) {
            Token = token;
            Prefix = prefix;
            MasterAdminId = masterAdminId;
            MasterGuildId = masterGuildId;
            BotLogChannels = botLogChannels;
            OtherAdmins = otherAdmins;
        }

        public void AssignMasterGuild(ulong guildId)
            => MasterGuildId = guildId;
        
        public bool AssignAdmin(ulong userId) {
            if (OtherAdmins.Contains(userId)) return false;

            OtherAdmins.Add(userId);
            return true;
        }

        public bool RemoveAdmin(ulong userId) {
            if (!OtherAdmins.Contains(userId)) return false;

            OtherAdmins.Remove(userId);
            return true;
        }

        public void AssignBotLogChannel(ulong channelId, BotLogType logType) {
            switch (logType) {
                case BotLogType.Common:
                    BotLogChannels.Common = channelId;
                    break;
                case BotLogType.CommandManager:
                    BotLogChannels.CommandMngr = channelId;
                    break;
                case BotLogType.GuildManager:
                    BotLogChannels.GuildMngr = channelId;
                    break;
                case BotLogType.MusicManager:
                    BotLogChannels.MusicMngr = channelId;
                    break;
                case BotLogType.MusicPlayerActive:
                    BotLogChannels.MusicPlayerActive = channelId;
                    break;
                case BotLogType.MusicPlayerError:
                    BotLogChannels.MusicPlayerError = channelId;
                    break;
                case BotLogType.AudioChord:
                    BotLogChannels.AudioChord = channelId;
                    break;
                case BotLogType.TaskScheduler:
                    BotLogChannels.TaskSch = channelId;
                    break;
                case BotLogType.TimerManager:
                    BotLogChannels.TimerMngr = channelId;
                    break;
            }
        }
    }
}